dataStructures/queues/queue.go

```diff
@@ -39,52 +39,52 @@ func copyQ(q Queue) Queue {
 	return append(Queue{}, q...)
 }
 
-// Add method takes an order and adds it to the queue based on priority.
-func (q *Queue) Add(order *Order) {
-	if len(*q) == 0 {
-		*q = Queue{order}
-	} else {
-		var appended bool
-
-	QLoop:
-		for i, existingOrder := range *q {
-			var insert int
-
-			switch {
-			case order.priority < existingOrder.priority:
-				insert = i
-			case order.priority == existingOrder.priority:
-				insert = 1 + i
-				// peek to check next order in the queue.
-				if insert < len(*q) && (*q)[insert].priority == order.priority {
-					// keep looking for the best location to insert.
-					continue QLoop
-				}
-			default:
-				continue QLoop
+// binarySearch method will help to pick a slot to fix the next order
+// with complexity of Olog(n)
+func getInsertionSlot(q Queue, value int) int {
+	low := 0
+	high := len(q) - 1
+
+	for low <= high {
+		mid := (high + low) / 2
+
+		switch {
+		case q[mid].priority == value:
+			if (mid+1) < len(q) && q[mid+1].priority == value {
+				low = mid + 1
+			} else {
+				return mid + 1
 			}
+		case q[mid].priority < value:
+			low = mid + 1
+		default:
+			high = mid - 1
```

Comment:
Are you sure it's not `low = mid` and `high = mid`? It's too early for me to think properly :stuck_out_tongue_closed_eyes:  but my gut tells me that you might accidentally miss things (argument against my gut is that you've checked it in the first case). There's a good chance I'm wrong!

This is why tests are most important and should be extremely simple to understand. Humans can only check code so much, so tests should catch the bugs and be very obvious about what they do.

Reply:
if `low = mid` and `high = mid`, I am heading for an infinite loop.

Reply:
What if you change your loop condition to `for low < high {`? Haven't you found your insertion point if it's `==`?

Reply:
If I change it to this, there will be a point that low will always be less than high and that will result in an infinite loop too.

I haven't found the insertion because I needed to handle the order with the same priorities.

Reply:
Ahh ok :) I forgot about the secondary ordering!